### PR TITLE
qv2ray: Add version 2.1.2

### PR DESCRIPTION
**Pull request by skt041959**
Yet another v2ray client. Qv2ray is a Qt based cross-platform GUI client using v2ray core.
More detail: https://github.com/Qv2ray/Qv2ray

### CODE REVIEW

File: bucket/qv2ray.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        [
+            "qv2ray.exe",
+            "qv2ray"
+        ]
+    ],
+    "checkver": "github",
+    "autoupdate": {
+        "url": "https://github.com/Qv2ray/Qv2ray/releases/download/$version/Qv2ray-refs.tags.$version-win64-gRPC.7z"
+    }
+}
```
Comment (Ash258) [inline bucket/qv2ray.json]:
```suggestion
{
    "version": "2.0.1",
    "description": "A Qt based cross-platform GUI client with v2ray core",
    "homepage": "https://github.com/Qv2ray/Qv2ray",
    "license": "GPL-3.0",
    "depends": "v2ray",
    "architecture": {
        "64bit": {
            "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v2.0.1/Qv2ray-refs.tags.v2.0.1-win64-gRPC.7z",
            "hash": "7b4427ebf8aa0bebc47f4dae5594a7fb9d92a381a182b9820f6eca64d122f009"
        }
    },
    "extract_dir": "release",
    "bin": "qv2ray.exe",
    "shortcuts": [
        [
            "qv2ray.exe",
            "qv2ray"
        ]
    ],
    "persist": "config",
    "checkver": "github",
    "autoupdate": {
        "architecture": {
            "64bit": {
                "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v$version/Qv2ray-refs.tags.v$version-win64-gRPC.7z"
            }
        }
    }
}

```

**Comment by skt041959** [comment]
@Ash258

> What is difference between gRPC and libqvb?

There will be only gRPC version in the future releases.

> This cannot work with default configuration

I will think about to specify the default v2ray core path installed by scoop 

I have removed the auto update section, because the developer will change release filename

**Comment by Ash258** [comment]
> There will be only gRPC version in the future releases.

What does it mean??

**Comment by skt041959** [comment]
> Source?
From the discussion with developers, they will only provide gRPC build for Windows from the next release.

File: bucket/qv2ray.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    ],
+    "checkver": {
+        "github": "https://github.com/Qv2ray/Qv2ray",
+        "regex": "/releases/tag/v([\\da-z.]+)"
+    },
+    "autoupdate": {
+        "architecture": {
+            "64bit": {
+                "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v$version/Qv2ray.v$version.Windows-x64.qt5.13.2.7z"
+            }
```
Comment (kidonng) [inline bucket/qv2ray.json]:
```diff
+ "32bit": {
+     "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v$version/Qv2ray.v$version.Windows-x86.qt5.13.2.7z"
+ }
```

File: bucket/qv2ray.json [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        ]
+    ],
+    "persist": "config",
+    "post_install": [
+        "cmd /c mklink /j \"$persist_dir\\config\\vcore\" \"$scoopdir\\apps\\v2ray\\current\""
+    ],
+    "checkver": {
+        "github": "https://github.com/Qv2ray/Qv2ray",
+        "regex": "/releases/tag/v([\\da-z.]+)"
+    },
```
Comment (kidonng) [inline bucket/qv2ray.json]:
```diff
- "checkver": {
-      "github": "https://github.com/Qv2ray/Qv2ray",
-      "regex": "/releases/tag/v([\\da-z.]+)"
-  },
+ "checkver": "github",
```

File: bucket/qv2ray.json [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    "license": "GPL-3.0",
+    "depends": [
+        "v2ray",
+        "vcredist2019"
+    ],
+    "architecture": {
+        "64bit": {
+            "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v2.1.2/Qv2ray.v2.1.2.Windows-x64.qt5.13.2.7z",
+            "hash": "14c209b27899d210d7c06a3fb827991855702f5d75922078f2b41ccae2c97915"
+        }
```
Comment (kidonng) [inline bucket/qv2ray.json]:
```diff
+ "32bit": {
+     "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v2.1.2/Qv2ray.v2.1.2.Windows-x86.qt5.13.2.7z",
+     "hash": "b1097d84bc4a7a9e1f0780788f6813757930e3fa50a3b845ced669cdd61db6ad"
+ }
```

File: bucket/qv2ray.json [hunk 5]
```diff
@@ -0,0 +1,39 @@
+{
+    "version": "2.1.2",
+    "description": "Graphical frontend for v2ray",
+    "homepage": "https://github.com/Qv2ray/Qv2ray",
+    "license": "GPL-3.0",
```
Comment (kidonng) [inline bucket/qv2ray.json]:
```diff
- "license": "GPL-3.0",
+ "license": "GPL-3.0-or-later",
```

File: bucket/qv2ray.json [hunk 6]
```diff
@@ -0,0 +1,39 @@
+{
+    "version": "2.1.2",
+    "description": "Graphical frontend for v2ray",
```
Comment (kidonng) [inline bucket/qv2ray.json]: I suggest using "Qv2ray is a cross-platform v2ray graphical frontend written in Qt." from https://github.com/Qv2ray/Qv2ray/wiki.

**Comment by kidonng** [comment]
Submitted an issue to Qv2ray asking for removing Qt version number from release files (linked above👆), which if accepted, should make autoupdate work even Qt version updates.

File: bucket/qv2ray.json [hunk 7]
```diff
@@ -0,0 +1,43 @@
+{
+    "version": "2.1.2",
+    "description": "A cross-platform v2ray graphical frontend written in Qt",
```
Comment (Ash258) [inline bucket/qv2ray.json]:
```suggestion
    "description": "Graphical frontend for v2ray",
```

File: bucket/qv2ray.json [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    "homepage": "https://github.com/Qv2ray/Qv2ray",
+    "license": "GPL-3.0-or-later",
```
Comment (Ash258) [inline bucket/qv2ray.json]:
```suggestion
    "license": "GPL-3.0-only",
```

File: bucket/qv2ray.json [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    "depends": [
+        "v2ray",
+        "vcredist2019"
+    ],
```
Comment (Ash258) [inline bucket/qv2ray.json]:
```suggestion
    "depends": "v2ray",
    "suggest": {
        "vcredist": "extras/vcredist2019"
    },
```

**Comment by kidonng** [comment]
> * Nowhere is specified `-or-later` license

https://github.com/Qv2ray/Qv2ray#licences

> * Noone care if it is cross platform or in what it is written

This is a brief description from [their wiki](https://github.com/Qv2ray/Qv2ray/wiki), which best describes what it is. There are many "Graphical frontend for v2ray", someone may care about consistent experience across different platforms, not "noone".

> * Source of vcredist requirement??

[v2.1.1 Release notes](https://github.com/Qv2ray/Qv2ray/releases/tag/v2.1.1).

> BREAKING: Windows 版本现在使用 MSVC 编译
> Windows 用户如果启动失败的话，请尝试安装 Qv2ray 安装包内的 MSVC 运行库

TL:

> BREAKING: Windows version is now compiled with MSVC
> Should Windows users fail to launch the program, please try installing MSVC runtime in Qv2ray's installation package.

**Comment by skt041959** [comment]
>     * Do not use unneeded custom symlinks when you can edit one file ...

What's the disadvantage to go with symlinks?

>     * Source of vcredist requirement??

The developer has bundled a vcredist2019 installer in their package. The vcredist2019 runtime library is a real dependency not suggestion.

File: bucket/qv2ray.json [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    "bin": "qv2ray.exe",
+    "shortcuts": [
+        [
+            "qv2ray.exe",
+            "qv2ray"
+        ]
+    ],
+    "persist": "config",
+    "post_install": [
+        "cmd /c mklink /j \"$persist_dir\\config\\vcore\" \"$scoopdir\\apps\\v2ray\\current\""
```
Comment (Ash258) [inline bucket/qv2ray.json]: Nonsense to use junction/sym link when you can use one line confifguration

Comment (Ash258) [reply bucket/qv2ray.json]: Remove vcredist installer from dir if presented

File: bucket/qv2ray.json [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+        "vcredist2019"
+    ],
+    "architecture": {
+        "64bit": {
+            "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v2.1.2/Qv2ray.v2.1.2.Windows-x64.qt5.13.2.7z",
+            "hash": "14c209b27899d210d7c06a3fb827991855702f5d75922078f2b41ccae2c97915"
+        },
+        "32bit": {
+            "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v2.1.2/Qv2ray.v2.1.2.Windows-x86.qt5.13.2.7z",
+            "hash": "b1097d84bc4a7a9e1f0780788f6813757930e3fa50a3b845ced669cdd61db6ad"
```
Comment (MaoKwen) [inline bucket/qv2ray.json]:
```suggestion
        "64bit": {
            "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v2.2.3/Qv2ray.v2.2.3.Windows-x64.7z",
            "hash": "9000fc88d29fc05b3d7052e6a6e1ee79e65ba2e65cb4f9ff42300035bb43e8e1"
        },
        "32bit": {
            "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v2.2.3/Qv2ray.v2.2.3.Windows-x86.7z",
            "hash": "dcf3df798094a200578d46e0bc3e8d084eeb813f54970f0523439e9522221f45"
```

File: bucket/qv2ray.json [hunk 12]
```diff
@@ -0,0 +1,43 @@
+{
+    "version": "2.1.2",
```
Comment (MaoKwen) [inline bucket/qv2ray.json]:
```suggestion
    "version": "2.2.3",
```
```suggestion
    "version": "2.1.2",
```

File: bucket/qv2ray.json [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        "cmd /c mklink /j \"$persist_dir\\config\\vcore\" \"$scoopdir\\apps\\v2ray\\current\""
+    ],
+    "checkver": "github",
+    "autoupdate": {
+        "architecture": {
+            "64bit": {
+                "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v$version/Qv2ray.v$version.Windows-x64.qt5.13.2.7z"
+            },
+            "32bit": {
+                "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v$version/Qv2ray.v$version.Windows-x86.qt5.13.2.7z"
```
Comment (MaoKwen) [inline bucket/qv2ray.json]:
```suggestion
            "64bit": {
                "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v$version/Qv2ray.v$version.Windows-x64.7z"
            },
            "32bit": {
                "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v$version/Qv2ray.v$version.Windows-x86.7z"
```

File: bucket/qv2ray.json [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    ],
+    "architecture": {
+        "64bit": {
+            "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v2.1.2/Qv2ray.v2.1.2.Windows-x64.qt5.13.2.7z",
+            "hash": "14c209b27899d210d7c06a3fb827991855702f5d75922078f2b41ccae2c97915"
+        },
+        "32bit": {
+            "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v2.1.2/Qv2ray.v2.1.2.Windows-x86.qt5.13.2.7z",
+            "hash": "b1097d84bc4a7a9e1f0780788f6813757930e3fa50a3b845ced669cdd61db6ad"
+        }
```
Comment (MaoKwen) [inline bucket/qv2ray.json]:
```suggestion
        "64bit": {
            "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v2.2.3/Qv2ray.v2.2.3.Windows-x64.7z",
            "hash": "9000fc88d29fc05b3d7052e6a6e1ee79e65ba2e65cb4f9ff42300035bb43e8e1"
        },
        "32bit": {
            "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v2.2.3/Qv2ray.v2.2.3.Windows-x86.7z,
            "hash": "dcf3df798094a200578d46e0bc3e8d084eeb813f54970f0523439e9522221f45"
        }
```

File: bucket/qv2ray.json [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    },
+    "extract_dir": "release",
```
Comment (MaoKwen) [inline bucket/qv2ray.json]:
```suggestion
    "extract_dir": "deployment",
```

File: bucket/qv2ray.json (hunk 14 above)
Comment (MaoKwen) [inline bucket/qv2ray.json]:
```suggestion
        "64bit": {
            "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v2.2.3/Qv2ray.v2.2.3.Windows-x64.7z",
            "hash": "9000fc88d29fc05b3d7052e6a6e1ee79e65ba2e65cb4f9ff42300035bb43e8e1"
        },
        "32bit": {
            "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v2.2.3/Qv2ray.v2.2.3.Windows-x86.7z,
            "hash": "dcf3df798094a200578d46e0bc3e8d084eeb813f54970f0523439e9522221f45"
        }
```

File: bucket/qv2ray.json (hunk 12 above)
Comment (MaoKwen) [inline bucket/qv2ray.json]:
```suggestion
    "version": "2.2.3",
```

File: bucket/qv2ray.json [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+    ],
+    "checkver": "github",
+    "autoupdate": {
+        "architecture": {
+            "64bit": {
+                "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v$version/Qv2ray.v$version.Windows-x64.qt5.13.2.7z"
+            },
+            "32bit": {
+                "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v$version/Qv2ray.v$version.Windows-x86.qt5.13.2.7z"
+            }
```
Comment (MaoKwen) [inline bucket/qv2ray.json]:
```suggestion
            "64bit": {
                "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v$version/Qv2ray.v$version.Windows-x64.7z"
            },
            "32bit": {
                "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v$version/Qv2ray.v$version.Windows-x86.7z"
            }
```

File: bucket/qv2ray.json [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+            "url": "https://github.com/Qv2ray/Qv2ray/releases/download/v2.1.2/Qv2ray.v2.1.2.Windows-x86.qt5.13.2.7z",
+            "hash": "b1097d84bc4a7a9e1f0780788f6813757930e3fa50a3b845ced669cdd61db6ad"
+        }
+    },
+    "extract_dir": "release",
+    "bin": "qv2ray.exe",
+    "shortcuts": [
+        [
+            "qv2ray.exe",
+            "qv2ray"
```
Comment (MaoKwen) [inline bucket/qv2ray.json]:
```suggestion
            "Qv2ray"
```

File: bucket/qv2ray.json [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+        ]
+    ],
+    "persist": "config",
+    "post_install": [
+        "cmd /c mklink /j \"$persist_dir\\config\\vcore\" \"$scoopdir\\apps\\v2ray\\current\""
+    ],
```
Comment (MaoKwen) [inline bucket/qv2ray.json]:
```suggestion
    "post_install": [
        "if (!(Test-Path \"$persist_dir\\config\\Qv2ray.conf\")) {",
        "    [PSObject]@{",
        "        config_version = 11",
        "        kernelConfig = [PSObject]@{",
        "            v2CorePath_win = \"$scoopdir\\apps\\v2ray\\current\\v2ray.exe\"",
        "            v2AssetsPath_win = \"$scoopdir\\apps\\v2ray\\current\"",
        "        }",
        "    } | ConvertTo-Json | Out-File -encoding ASCII \"$persist_dir\\config\\Qv2ray.conf\"",
        "}"
    ],
```

**Comment by DuckSoft** [comment]
Hello guys, I wonder if this is ready to go.

File: bucket/qv2ray.json [hunk 19]
```diff
@@ -0,0 +1,53 @@
+{
+    "version": "2.2.3",
```
Comment (DuckSoft) [inline bucket/qv2ray.json]: Qv2ray has updated to [`2.3.1`](https://github.com/Qv2ray/Qv2ray/releases/v2.3.1).
We shall consider updating this.

**Comment by Ash258** [comment]
@DuckSoft Will be updated automatically after merge. Or i will update it before merging next week.

**Comment by DuckSoft** [comment]
> @DuckSoft Will be updated automatically after merge. Or i will update it before merging next week.

An automatic update shall be fine. Thanks for your efforts btw!

**Pull request merged**